File: Binary Tree/111.二叉树的最小深度.cpp
using namespace std;
/*
 * @lc app=leetcode.cn id=111 lang=cpp
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int getLevel(TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }
        int leftDepth = getLevel(root->left);   // 左
        int rightDepth = getLevel(root->right); // 右
                                                // 中
        // 当一个左子树为空，右不为空，这时并不是最低点
        if (root->left == nullptr && root->right != nullptr)
        {
            return 1 + rightDepth;
        }
        // 当一个右子树为空，左不为空，这时并不是最低点
        if (root->left != nullptr && root->right == nullptr)
        {
            return 1 + leftDepth;
        }

        int result = 1 + min(leftDepth, rightDepth);
        return result;
    }

    int minDepth(TreeNode *root)
    {
        return getLevel(root);
    }
};
// @lc code=end

File: Binary Tree/二叉树的属性/111.二叉树的最小深度.ts
/*
 * @lc app=leetcode.cn id=111 lang=typescript
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function minDepth(root: TreeNode | null): number {
  let queue = [root]
  let level = 0

  if (!root) {
    return level
  }

  while (queue.length > 0) {
    // 当前层级的左右节点数量
    let size = queue.length

    // 移除当前层级节点，移除出的同时将子节点放入队列
    for (let i = 0; i < size; i++) {
      let node = queue.shift()

      if (node) {
        // 当前即为最小深度
        if (!node.left && !node.right) {
          return ++level
        }

        node.left && queue.push(node.left)
        node.right && queue.push(node.right)
      }
    }
    // 遍历完一层，层级+1
    level++
  }

  return level
};
// @lc code=end

